src/event-stream/tokens-contract-handler.ts

```diff
@@ -626,27 +643,77 @@ class TokensContractHandler {
   /**
    * Make readonly contract call
    */
-  private async makeReadOnlyContractCall(
+  async makeReadOnlyContractCall(
+    functionName: string,
+    functionArgs: ClarityValue[]
+  ): Promise<ReadOnlyContractCallResponse> {
+    const url = this.stacksNetwork.getReadOnlyFunctionCallApiUrl(
+      this.contractAddress,
+      this.contractName,
+      functionName
+    );
+    const args = functionArgs.map(arg => cvToHex(arg));
+
+    const body = JSON.stringify({
+      sender: this.address,
+      arguments: args,
+    });
+
+    const response = await fetchPrivate(url, {
+      method: 'POST',
+      body,
+      headers: {
+        'Content-Type': 'application/json',
+      },
+    });
+
+    if (!response.ok) {
+      if (response.status >= 500) return { found: false, retriable: true };
+      const msg = await response.text().catch(() => '');
+      return {
+        found: false,
+        retriable: false,
+        error: `Error calling read-only function. Response ${response.status}: ${response.statusText}. Attempted to fetch ${url} and failed with the message: "${msg}"`,
+      };
+    }
+    const value = await response.json();
+    return { found: true, value: parseReadOnlyResponse(value) };
+  }
+
+  async retryReadOnlyContractCall(
     functionName: string,
     functionArgs: ClarityValue[]
   ): Promise<ClarityValue> {
-    const txOptions: ReadOnlyFunctionOptions = {
-      senderAddress: this.address,
-      contractAddress: this.contractAddress,
-      contractName: this.contractName,
-      functionName: functionName,
-      functionArgs: functionArgs,
-      network: this.stacksNetwork,
-    };
-    return await callReadOnlyFunction(txOptions);
+    return await new Promise<ClarityValue>(async (resolve, reject) => {
+      let response: ReadOnlyContractCallResponse = { found: false, retriable: true };
+      if (isMetadataStrictModeEnabled()) {
+        while (!response.found && response.retriable) {
+          response = await this.makeReadOnlyContractCall(functionName, functionArgs);
+        }
+        if (response.found) return resolve(response.value);
+        // incase the issue is not temporary.
+        else if (!response.retriable) return reject(new Error(response.error));
+      } else {
+        const interval = setInterval(async () => {
```

Comment:
The `setInterval` usage here is a bit confusing to reason about. It looks like essentially a while-loop in the form of callbacks, right? In that case, just a regular while loop with an `await timeout(interval)` would be a lot more clear.